### PR TITLE
Implement transparent client pattern and middleware error handling

### LINKED ISSUE #137: Implement transparent client pattern and middleware error handling
Not in scope for this PR, as it would make it too large and this can focus on introducing the client into the existing module, but we should follow a transparent pattern for using the client [as was done for Python](https://github.com/seamapi/python/blob/cb26ee4f9544fafae3e3614f1b02eda41a3445a6/seam/routes/access_codes_simulate.py#L23-L27) and move error handling into middleware.

_Originally posted by @razor-x in https://github.com/seamapi/ruby-next/pull/134#discussion_r1811989551_

### PR DESCRIPTION
**Pull request by andrii-balitskyi**
Closes https://github.com/seamapi/ruby-next/issues/137

- **Add response middleware**
- **Remove base client**
- **Remove request_seam methods in favour of using faraday client directly**
- **Support [] syntax on DeepHashAccessor instances**
- **Adjust action attempt polling to the new logic**
- **Bump the generator to 1.14.10 to adjust routes**
- **Fix tests**

### CODE REVIEW

File: lib/seam/helpers/action_attempt.rb [hunk 1]
```diff
@@ -37,13 +35,7 @@ def self.wait_until_finished(action_attempt, client, timeout: nil, polling_inter
...
 
       def self.update_action_attempt(action_attempt, client)
-        response = client.request_seam(
-          :post,
-          "/action_attempts/get",
-          body: {
-            action_attempt_id: action_attempt.action_attempt_id
-          }
-        )
+        response = client.post("/action_attempts/get", {action_attempt_id: action_attempt.action_attempt_id})
```
Comment (razor-x) [inline lib/seam/helpers/action_attempt.rb:38]:
```suggestion
        response = client.get("/action_attempts/get", {action_attempt_id: action_attempt.action_attempt_id})
```

File: lib/seam/http_single_workspace.rb [hunk 2]
```diff
@@ -18,9 +18,10 @@ def initialize(client: nil, api_key: nil, personal_access_token: nil, workspace_
         @endpoint = options[:endpoint]
         @auth_headers = options[:auth_headers]
         @defaults = Seam::DeepHashAccessor.new({"wait_for_action_attempt" => wait_for_action_attempt})
-
         @client = client || Seam::Http::Request.create_faraday_client(@endpoint, @auth_headers, faraday_options,
           faraday_retry_options)
+
+        initialize_routes(client: @client, defaults: @defaults)
```
Comment (razor-x) [inline lib/seam/http_single_workspace.rb:24]: What does this do? Why is it defined as a private function but called here?

File: lib/seam/request.rb [hunk 3]
```diff
@@ -72,6 +39,48 @@ def self.default_headers
...
+            raise Faraday::RetriableResponse.new(nil, env.response)
+          end
+
+          raise Http::UnauthorizedError.new(request_id) if status_code == 401
+
+          body = begin
+            JSON.parse(env.body)
+          rescue
+            {}
+          end
```
Comment (razor-x) [inline lib/seam/request.rb:65]: We can't assume the response is always JSON. It's possible for things in front of the API to return errors that are not JSON like Bad Gateway or Service Unavailable. I checked and the python SDK has the same assumption: https://github.com/seamapi/python/blob/7b62fa7eab7094d36af6f74465d4703da94d4330/seam/client.py#L64


We may need to replicate this logic in all the SDKs: https://github.com/seamapi/javascript-http/blob/8178e4f42e89198defdcda0a83a7b6ed4b482066/src/lib/seam/connect/error-interceptor.ts#L34-L61

I'm not sure we should keep the default "unknown" values either: the error needs to have a `type` and a `message` to be considered a Seam error. 

Basically:

1. Check for 401 status code and throw our custom SeamHttpUnauthorizedError
2. Try to determine if the error response can reasonably assumed to be a JSON error from the Seam API and throw it as either a `SeamHttpInvalidInputError` or the generic `SeamHttpApiError`. We can still default  
3. Otherwise, we just throw the error provided by the underlying http client.

File: lib/seam/http_single_workspace.rb (hunk 2 above)
Comment (andrii-balitskyi) [reply lib/seam/http_single_workspace.rb:24]: `initialize_routes` is an internal method to set up route clients by making available `@client` and `@defaults` for them. `SingleWorkspace` can access private `initialize_routes` because `Routes` module gets `include`d in the class. While the method will exist on instances, being private it can only be called internally (inside the class/module methods) and not from outside code, unlike public methods like 'devices', 'access_codes' etc. which can be called by anyone using the instance.

File: lib/seam/request.rb (hunk 3 above)
Comment (andrii-balitskyi) [reply lib/seam/request.rb]: b35a6b17c89dd35e50de60682dd1de7d9c501457

File: lib/seam/http_single_workspace.rb (hunk 2 above)
Comment (razor-x) [reply lib/seam/http_single_workspace.rb:24]: Why is this method needed though vs just putting the code inline? I think I must be missing something obvious.

File: lib/seam/request.rb [hunk 4]
```diff
@@ -49,35 +49,51 @@ def on_complete(env)
           return if env.success?
 
           status_code = env.status
-          request_id = env.response_headers["seam-request-id"]
 
           retry_statuses = @retry_options.fetch(:retry_statuses, [])
           if retry_statuses.include?(status_code)
             raise Faraday::RetriableResponse.new(nil, env.response)
```
Comment (razor-x) [inline lib/seam/request.rb:55]: I'm surprised we have to do this manually vs the retry middleware does this first before the Seam error handler.

**Review by razor-x — APPROVED** [review]

File: lib/seam/request.rb (hunk 4 above)
Comment (andrii-balitskyi) [reply lib/seam/request.rb:55]: Oh yeah, I hadn't thought about this. I think we can achieve that by putting the retry middleware last in the Faraday connection builder, since the last middleware attached is the first to process the response. Nice catch!

Comment (andrii-balitskyi) [reply lib/seam/request.rb]: bbcf4e5062f4317f3c688c05ab93e837cd6d6c3d

File: lib/seam/http_single_workspace.rb (hunk 2 above)
Comment (razor-x) [reply lib/seam/http_single_workspace.rb:24]: I was wondering why not to just do
```
      @client = client
      @defaults = defaults
``` 

But now I understand what this is doing. Basically this module only works if mixed into classes that define `@client` and `@defaults`. And this method is helping clarify that requirement.

I wonder if there is a more standard pattern to handle this.
I think it may be what this question is asking about https://stackoverflow.com/questions/12586051/initializing-instance-variables-in-mixins

**Review by razor-x — APPROVED** [review]
